src/Interpreters/Cache/QueryResultCache.cpp

```diff
@@ -0,0 +1,343 @@
+            cache_size_in_bytes -= it->second.allocatedBytes(); /// key replacement
+
+        /// cache[key] = query_result; /// does no replacement for unclear reasons
+        cache.erase(key);
+        cache[key] = std::move(query_result);
+
+        LOG_DEBUG(&Poco::Logger::get("QueryResultCache"), "Stored result of query {}", key.queryStringFromAst());
+    }
+}
+
+void QueryResultCache::Writer::buffer(Chunk && partial_query_result)
+{
+    if (skip_insert)
+        return;
+
+    partial_query_results.emplace_back(std::move(partial_query_result));
+
+    new_entry_size_in_bytes += partial_query_results.back().allocatedBytes();
+    new_entry_size_in_rows += partial_query_results.back().getNumRows();
+
+    if ((new_entry_size_in_bytes > max_entry_size_in_bytes) || (new_entry_size_in_rows > max_entry_size_in_rows))
+        skip_insert = true;
+}
+
+QueryResultCache::Reader::Reader(const Cache & cache_, const Key & key, size_t & cache_size_in_bytes_)
+{
+    auto it = cache_.find(key);
+
+    if (it == cache_.end())
+    {
+        LOG_DEBUG(&Poco::Logger::get("QueryResultCache"), "No entry found for query {}", key.queryStringFromAst());
+        return;
+    }
+
+    if (it->first.username.has_value() && it->first.username != key.username)
+    {
+        LOG_DEBUG(&Poco::Logger::get("QueryResultCache"), "Inaccessible entry found for query {}", key.queryStringFromAst());
+        return;
+    }
+
+    if (is_stale(it->first))
+    {
+        cache_size_in_bytes_ -= it->second.allocatedBytes();
+        const_cast<Cache &>(cache_).erase(it);
+        LOG_DEBUG(&Poco::Logger::get("QueryResultCache"), "Stale entry found and removed for query {}", key.queryStringFromAst());
+        return;
+    }
+
+    LOG_DEBUG(&Poco::Logger::get("QueryResultCache"), "Entry found for query {}", key.queryStringFromAst());
+    pipe = Pipe(std::make_shared<SourceFromSingleChunk>(key.header, it->second.clone()));
+}
+
+bool QueryResultCache::Reader::hasCacheEntryForKey() const
+{
+    bool res = !pipe.empty();
+
+    if (res)
+        ProfileEvents::increment(ProfileEvents::QueryResultCacheHits);
+    else
+        ProfileEvents::increment(ProfileEvents::QueryResultCacheMisses);
+
+    return res;
+}
+
+Pipe && QueryResultCache::Reader::getPipe()
+{
+    assert(!pipe.empty()); // cf. hasCacheEntryForKey()
+    return std::move(pipe);
+}
+
+QueryResultCache::QueryResultCache(size_t max_cache_size_in_bytes_, size_t max_cache_entries_, size_t max_cache_entry_size_in_bytes_, size_t max_cache_entry_size_in_rows_)
+    : max_cache_size_in_bytes(max_cache_size_in_bytes_)
+    , max_cache_entries(max_cache_entries_)
+    , max_cache_entry_size_in_bytes(max_cache_entry_size_in_bytes_)
+    , max_cache_entry_size_in_rows(max_cache_entry_size_in_rows_)
+{
+}
+
+QueryResultCache::Reader QueryResultCache::createReader(const Key & key)
+{
+    std::lock_guard lock(mutex);
+    return Reader(cache, key, cache_size_in_bytes);
```

Comment:
minor, but imo would be better to pass `lock` here as an argument.

Reply:
Done, tnx. I found TSA does not play well with subclasses :-(